File: sdks/js/langgraph/cases/2-multi-step.js
/**
 * 2-multi-step: Multi-step Conversation
 *
 * Tests a multi-step conversation with conversation history using LangGraph SDK
 * and verifies that Sentry captures all spans for both API calls.
 */

const { Sentry } = require("../setup");
const { ChatOpenAI } = require("@langchain/openai");
const { createReactAgent } = require("@langchain/langgraph/prebuilt");
const {
  HumanMessage,
  SystemMessage,
  AIMessage,
} = require("@langchain/core/messages");
const { runTestCase } = require("../../_test-utils/test-runner.cjs");

async function testLogic(inputs) {
  const { model, system, first_prompt, second_prompt } = inputs;

  // Create LLM instance
  const llm = new ChatOpenAI({
    modelName: model,
    apiKey: process.env.OPENAI_API_KEY,
  });

  // Create a simple react agent with no tools
  const agent = createReactAgent({ llm, tools: [] });

  // First call
  const firstResult = await agent.invoke({
    messages: [new SystemMessage(system), new HumanMessage(first_prompt)],
  });

  const firstMessages = firstResult.messages;
  const firstLastMessage = firstMessages[firstMessages.length - 1];
  const firstText = firstLastMessage.content;

  if (!firstText) {
    throw new Error("No completion returned from LangGraph (first call)");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    First response: ${firstText}`);
  }

  // Second call with conversation history
  const secondResult = await agent.invoke({
    messages: [
      new SystemMessage(system),
      new HumanMessage(first_prompt),
      new AIMessage(firstText),
      new HumanMessage(second_prompt),
    ],
  });

  const secondMessages = secondResult.messages;
  const secondLastMessage = secondMessages[secondMessages.length - 1];
  const secondText = secondLastMessage.content;

  if (!secondText) {
    throw new Error("No completion returned from LangGraph (second call)");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    Second response: ${secondText}`);
  }
}

module.exports = runTestCase("2-multi-step", testLogic, Sentry);

File: sdks/py/langgraph/cases/2-multi-step.py
"""
2-multi-step: Multi-step Conversation

Tests a multi-step conversation with conversation history using LangGraph SDK
and verifies that Sentry captures all spans for both API calls.
"""

import os
from langgraph.prebuilt import create_react_agent
from langchain_openai import ChatOpenAI
from test_runner import run_test_case


async def test_logic(inputs):
    """The actual test logic"""
    model = inputs["model"]
    system = inputs["system"]
    first_prompt = inputs["first_prompt"]
    second_prompt = inputs["second_prompt"]

    llm = ChatOpenAI(model=model, api_key=os.getenv("OPENAI_API_KEY"))

    # Create a simple react agent with no tools
    agent = create_react_agent(llm, tools=[])

    # First call
    first_messages = [
        ("system", system),
        ("user", first_prompt)
    ]

    first_result = agent.invoke({"messages": first_messages})
    first_text = first_result["messages"][-1].content

    if not first_text:
        raise Exception("No output returned from LangGraph (first call)")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    First response: {first_text}")

    # Second call with conversation history
    second_messages = [
        ("system", system),
        ("user", first_prompt),
        ("ai", first_text),
        ("user", second_prompt)
    ]

    second_result = agent.invoke({"messages": second_messages})
    second_text = second_result["messages"][-1].content

    if not second_text:
        raise Exception("No output returned from LangGraph (second call)")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    Second response: {second_text}")

    return second_text


# Export test case functions
test_case = run_test_case("2-multi-step", test_logic)
main = test_case["main"]
assert_sentry = test_case["assert_sentry"]
